File: src/frontend/src/api/apoliceSegurado.ts
import { clienteApi } from './clienteHttp'
import type { components } from './tipos-gerados'

/** Os dados da apólice exibíveis ao segurado ativo. */
export type ApoliceSegurado = {
  numero: string
  tipo: string
  situacao: string
  estadoObjetivo: string
  vigenciaInicio: string
  vigenciaFim: string
  enderecoRiscoSintetico: string
  coberturas: string[]
  canalPreferido: string
  participaDeAlertas: boolean
}

/** Um critério comparado pela regra determinística, relevante à apólice. */
export type CriterioApolice = {
  operando: string
  valorObservado: string
  atende: boolean
  justificativa: string
}

/** A explicação de critérios relevantes à apólice, do snapshot de uma execução. */
export type ExplicacaoApolice = {
  elegibilidadeId: string
  criterios: CriterioApolice[]
}

type CorpoProblema = Partial<components['schemas']['ProblemaApoliceSegurado']>

type DadosErroApoliceSegurado = {
  codigo: string
  correlacaoId: string | null
  ocorrencia: string
  impacto: string
  proximaAcao: string
  status: number | null
}

/** Falha tipada da apólice/explicação, com ocorrência, impacto e próxima ação segura. */
export class ErroApoliceSegurado extends Error {
  readonly codigo: string
  readonly correlacaoId: string | null
  readonly ocorrencia: string
  readonly impacto: string
  readonly proximaAcao: string
  readonly status: number | null

  constructor(dados: DadosErroApoliceSegurado) {
    super(dados.ocorrencia)
    this.name = 'ErroApoliceSegurado'
    this.codigo = dados.codigo
    this.correlacaoId = dados.correlacaoId
    this.ocorrencia = dados.ocorrencia
    this.impacto = dados.impacto
    this.proximaAcao = dados.proximaAcao
    this.status = dados.status
  }
}

const FALHA_INESPERADA = {
  ocorrencia: 'A apólice não pôde ser consultada.',
  impacto: 'Os dados da apólice podem estar desatualizados ou indisponíveis.',
  proximaAcao: 'Verifique se o backend local está disponível e tente novamente.',
}

const FALHA_DE_REDE: DadosErroApoliceSegurado = {
  codigo: 'falha_de_rede',
  correlacaoId: null,
  ocorrencia: 'Não foi possível falar com o backend local da Central Preventiva.',
  impacto: 'Nenhum dado de apólice pôde ser exibido.',
  proximaAcao: 'Confirme que o backend está em execução em 127.0.0.1:8000 e tente de novo.',
  status: null,
}

function erroDeResultado(erro: unknown, status: number): ErroApoliceSegurado {
  const corpo = (erro ?? {}) as CorpoProblema
  return new ErroApoliceSegurado({
    codigo: corpo.codigo ?? 'falha_inesperada',
    correlacaoId: corpo.correlacao_id ?? null,
    ocorrencia: corpo.ocorrencia ?? FALHA_INESPERADA.ocorrencia,
    impacto: corpo.impacto ?? FALHA_INESPERADA.impacto,
    proximaAcao: corpo.proxima_acao ?? FALHA_INESPERADA.proximaAcao,
    status,
  })
}

function paraApolice(corpo: components['schemas']['RespostaApolice']): ApoliceSegurado {
  return {
    numero: corpo.numero,
    tipo: corpo.tipo,
    situacao: corpo.situacao,
    estadoObjetivo: corpo.estado_objetivo,
    vigenciaInicio: corpo.vigencia_inicio,
    vigenciaFim: corpo.vigencia_fim,
    enderecoRiscoSintetico: corpo.endereco_risco_sintetico,
    coberturas: corpo.coberturas,
    canalPreferido: corpo.canal_preferido,
    participaDeAlertas: corpo.participa_de_alertas,
  }
}

function paraCriterio(
  corpo: components['schemas']['RespostaCriterioApolice'],
): CriterioApolice {
  return {
    operando: corpo.operando,
    valorObservado: corpo.valor_observado,
    atende: corpo.atende,
    justificativa: corpo.justificativa,
  }
}

function paraExplicacao(
  corpo: components['schemas']['RespostaExplicacaoApolice'],
): ExplicacaoApolice {
  return {
    elegibilidadeId: corpo.elegibilidade_id,
    criterios: corpo.criterios.map(paraCriterio),
  }
}

async function chamarApolice(seguradoId: string) {
  return clienteApi.GET('/api/v1/segurados/{segurado_id}/apolice', {
    params: { path: { segurado_id: seguradoId } },
    fetch,
  })
}

async function chamarExplicacao(seguradoId: string, elegibilidadeId: string) {
  return clienteApi.GET(
    '/api/v1/segurados/{segurado_id}/apolice/explicacao/{elegibilidade_id}',
    {
      params: { path: { segurado_id: seguradoId, elegibilidade_id: elegibilidadeId } },
      fetch,
    },
  )
}

/** Consulta a apólice do segurado, ou lança `ErroApoliceSegurado` (`status` 404 se não
 * houver apólice ou não pertencer ao segurado). */
export async function getApolice(seguradoId: string): Promise<ApoliceSegurado> {
  let resultado: Awaited<ReturnType<typeof chamarApolice>>
  try {
    resultado = await chamarApolice(seguradoId)
  } catch {
    throw new ErroApoliceSegurado(FALHA_DE_REDE)
  }

  if (!resultado.response.ok || !resultado.data) {
    throw erroDeResultado(resultado.error, resultado.response.status)
  }

  return paraApolice(resultado.data)
}

/** Consulta a explicação de critérios de uma execução, ou lança `ErroApoliceSegurado`
 * (`status` 404 se a elegibilidade não existir ou não pertencer ao segurado). */
export async function getExplicacaoApolice(
  seguradoId: string,
  elegibilidadeId: string,
): Promise<ExplicacaoApolice> {
  let resultado: Awaited<ReturnType<typeof chamarExplicacao>>
  try {
    resultado = await chamarExplicacao(seguradoId, elegibilidadeId)
  } catch {
    throw new ErroApoliceSegurado(FALHA_DE_REDE)
  }

  if (!resultado.response.ok || !resultado.data) {
    throw erroDeResultado(resultado.error, resultado.response.status)
  }

  return paraExplicacao(resultado.data)
}

File: src/backend/central_preventiva/adaptadores/http/apolice_segurado.py
"""Recurso REST/JSON da apólice do segurado ativo e da explicação de critérios
(APOLICE-01..06, 5.3).

`GET /segurados/{segurado_id}/apolice` devolve os dados cadastrais e o estado objetivo da
apólice. `GET /segurados/{segurado_id}/apolice/explicacao/{elegibilidade_id}` devolve a
explicação de critérios de uma execução específica — uma apólice/elegibilidade inexistente
e uma de outro segurado devolvem exatamente o mesmo `404` genérico (APOLICE-04, AD-011),
mesmo padrão de 4.2/4.3/5.2.
"""

from uuid import UUID, uuid4

from fastapi import APIRouter
from fastapi.responses import JSONResponse
from pydantic import BaseModel, ConfigDict, Field

from central_preventiva.adaptadores.persistencia.repositorio_apolices import (
    RepositorioApolices,
)
from central_preventiva.adaptadores.persistencia.repositorio_elegibilidade import (
    RepositorioElegibilidades,
)
from central_preventiva.adaptadores.persistencia.repositorio_segurados import (
    RepositorioSegurados,
)
from central_preventiva.aplicacao.apolice_segurado import (
    ApoliceSegurado,
    ExplicacaoApolice,
    PortasApoliceSegurado,
    ServicoApoliceSegurado,
)
from central_preventiva.composicao.configuracao import Configuracao
from central_preventiva.dominio.avaliador_risco import Criterio

TIPO_PROBLEMA = "application/problem+json"
CAMINHO_APOLICE = "/segurados/{segurado_id}/apolice"
CAMINHO_EXPLICACAO = "/segurados/{segurado_id}/apolice/explicacao/{elegibilidade_id}"


class RespostaApolice(BaseModel):
    """Os dados da apólice exibíveis ao segurado ativo."""

    model_config = ConfigDict(extra="forbid")

    numero: str = Field(description="Número da apólice.")
    tipo: str = Field(description="Tipo da apólice (`residencial`/`automovel`).")
    situacao: str = Field(description="Situação cadastral (`ativa`/`cancelada`/`suspensa`).")
    estado_objetivo: str = Field(
        description="Estado objetivo textual — inclui `expirada` quando a vigência já "
        "passou, mesmo com `situacao='ativa'`. Nunca uma falha técnica."
    )
    vigencia_inicio: str = Field(description="Data de início de vigência (ISO 8601).")
    vigencia_fim: str = Field(description="Data de fim de vigência (ISO 8601).")
    endereco_risco_sintetico: str = Field(description="Endereço sintético do risco.")
    coberturas: list[str] = Field(description="Coberturas contratadas.")
    canal_preferido: str = Field(description="Canal de comunicação preferencial.")
    participa_de_alertas: bool = Field(description="Se o segurado participa de alertas.")


class RespostaCriterioApolice(BaseModel):
    """Um critério comparado pela regra determinística, relevante à apólice."""

    model_config = ConfigDict(extra="forbid")

    operando: str = Field(description="O que foi comparado.")
    valor_observado: str = Field(description="O valor observado no momento da avaliação.")
    atende: bool = Field(description="Se o critério foi atendido.")
    justificativa: str = Field(description="Explicação da comparação, sem promessa de cobertura.")


class RespostaExplicacaoApolice(BaseModel):
    """A explicação de critérios relevantes à apólice, do snapshot de uma execução."""

    model_config = ConfigDict(extra="forbid")

    elegibilidade_id: UUID = Field(description="Identificador da elegibilidade de origem.")
    criterios: list[RespostaCriterioApolice] = Field(
        description="Critérios relevantes à apólice (área, tipo, situação, cobertura)."
    )


class ProblemaApoliceSegurado(BaseModel):
    """Falha da apólice/explicação, com ocorrência, impacto e próxima ação segura."""

    model_config = ConfigDict(extra="forbid")

    codigo: str = Field(description="Código estável que identifica o tipo da falha.")
    correlacao_id: str = Field(description="Identificador único desta ocorrência de falha.")
    ocorrencia: str = Field(description="O que aconteceu, em português brasileiro.")
    impacto: str = Field(description="Efeito prático da falha para quem consultou o recurso.")
    proxima_acao: str = Field(description="Próxima ação segura recomendada para contornar a falha.")


def _problema(
    status: int, codigo: str, ocorrencia: str, impacto: str, proxima_acao: str
) -> JSONResponse:
    """Monta a resposta `problem+json` correlacionada de uma falha."""

    corpo = ProblemaApoliceSegurado(
        codigo=codigo,
        correlacao_id=str(uuid4()),
        ocorrencia=ocorrencia,
        impacto=impacto,
        proxima_acao=proxima_acao,
    )
    return JSONResponse(status_code=status, media_type=TIPO_PROBLEMA, content=corpo.model_dump())


def _problema_identificador_invalido() -> JSONResponse:
    """Resposta de identificador que não é um UUID válido."""

    return _problema(
        422,
        "identificador_invalido",
        "O identificador do segurado ou da elegibilidade não é um UUID válido.",
        "Nenhum dado pode ser exibido.",
        "Consulte pelos identificadores UUID retornados pela API.",
    )


def _problema_apolice_nao_encontrada() -> JSONResponse:
    """Resposta idêntica para apólice inexistente e apólice de outro segurado (AD-011)."""

    return _problema(
        404,
        "apolice_nao_encontrada",
        "Não existe apólice para este segurado.",
        "Nenhum dado de apólice pode ser exibido.",
        "Consulte a apólice pelo identificador do segurado ativo.",
    )


def _problema_explicacao_nao_encontrada(elegibilidade_id: str) -> JSONResponse:
    """Resposta idêntica para explicação inexistente e de outro segurado (AD-011)."""

    return _problema(
        404,
        "explicacao_nao_encontrada",
        f"A explicação '{elegibilidade_id}' não existe para este segurado.",
        "Nenhuma explicação pode ser exibida.",
        "Consulte a explicação pelo identificador de elegibilidade retornado pela API.",
    )


def _resposta_apolice(apolice: ApoliceSegurado) -> RespostaApolice:
    """Traduz a apólice resolvida pelo caso de uso para o contrato público."""

    return RespostaApolice(
        numero=apolice.numero,
        tipo=apolice.tipo,
        situacao=apolice.situacao,
        estado_objetivo=apolice.estado_objetivo,
        vigencia_inicio=apolice.vigencia_inicio.isoformat(),
        vigencia_fim=apolice.vigencia_fim.isoformat(),
        endereco_risco_sintetico=apolice.endereco_risco_sintetico,
        coberturas=list(apolice.coberturas),
        canal_preferido=apolice.canal_preferido,
        participa_de_alertas=apolice.participa_de_alertas,
    )


def _resposta_criterio(criterio: Criterio) -> RespostaCriterioApolice:
    """Traduz um critério para o contrato público."""

    return RespostaCriterioApolice(
        operando=criterio.operando,
        valor_observado=criterio.valor_observado,
        atende=criterio.atende,
        justificativa=criterio.justificativa,
    )


def _resposta_explicacao(explicacao: ExplicacaoApolice) -> RespostaExplicacaoApolice:
    """Traduz a explicação resolvida pelo caso de uso para o contrato público."""

    return RespostaExplicacaoApolice(
        elegibilidade_id=explicacao.elegibilidade_id,
        criterios=[_resposta_criterio(criterio) for criterio in explicacao.criterios],
    )


def montar_servico_apolice_segurado(configuracao: Configuracao) -> ServicoApoliceSegurado:
    """Compõe o serviço da apólice sobre os repositórios reais do DuckDB local."""

    caminho = configuracao.caminho_banco
    return ServicoApoliceSegurado(
        PortasApoliceSegurado(
            apolices=RepositorioApolices(caminho),
            segurados=RepositorioSegurados(caminho),
            elegibilidades=RepositorioElegibilidades(caminho),
        )
    )


def criar_roteador(configuracao: Configuracao) -> APIRouter:
    """Compõe o recurso da apólice e da explicação de critérios do segurado ativo."""

    roteador = APIRouter(tags=["Apólice do segurado"])
    servico = montar_servico_apolice_segurado(configuracao)

    @roteador.get(
        CAMINHO_APOLICE,
        response_model=RespostaApolice,
        status_code=200,
        summary="Consultar a apólice do segurado ativo",
        description=(
            "Devolve número, situação, vigência, endereço sintético do risco, "
            "coberturas, canal preferencial e participação em alertas. O estado "
            "objetivo distingue `expirada` de `cancelada`/`suspensa`, nunca como falha."
        ),
        responses={
            200: {"description": "Apólice encontrada para este segurado."},
            404: {
                "description": "Segurado sem apólice, ou apólice de outro segurado.",
                "model": ProblemaApoliceSegurado,
            },
            422: {
                "description": "O identificador do segurado não é um UUID válido.",
                "model": ProblemaApoliceSegurado,
            },
        },
    )
    async def consultar_apolice(segurado_id: str) -> RespostaApolice | JSONResponse:  # pyright: ignore[reportUnusedFunction]
        """Traduz `ServicoApoliceSegurado.obter` para o contrato público."""

        try:
            segurado_uuid = UUID(segurado_id)
        except ValueError:
            return _problema_identificador_invalido()

        apolice = servico.obter(segurado_uuid)
        if apolice is None:
            return _problema_apolice_nao_encontrada()
        return _resposta_apolice(apolice)

    @roteador.get(
        CAMINHO_EXPLICACAO,
        response_model=RespostaExplicacaoApolice,
        status_code=200,
        summary="Consultar a explicação de critérios de uma execução",
        description=(
            "Devolve os critérios relevantes à apólice (área, tipo, situação, "
            "cobertura) comparados pela regra determinística numa execução específica — "
            "sempre o snapshot da execução, nunca a apólice atual. Uma elegibilidade "
            "inexistente e uma de outro segurado devolvem exatamente a mesma resposta "
            "`404`."
        ),
        responses={
            200: {"description": "Explicação encontrada para este segurado."},
            404: {
                "description": "Elegibilidade inexistente ou de outro segurado.",
                "model": ProblemaApoliceSegurado,
            },
            422: {
                "description": "Algum identificador não é um UUID válido.",
                "model": ProblemaApoliceSegurado,
            },
        },
    )
    async def consultar_explicacao(  # pyright: ignore[reportUnusedFunction]
        segurado_id: str, elegibilidade_id: str
    ) -> RespostaExplicacaoApolice | JSONResponse:
        """Traduz `ServicoApoliceSegurado.obter_explicacao` para o contrato público."""

        try:
            segurado_uuid = UUID(segurado_id)
            elegibilidade_uuid = UUID(elegibilidade_id)
        except ValueError:
            return _problema_identificador_invalido()

        explicacao = servico.obter_explicacao(segurado_uuid, elegibilidade_uuid)
        if explicacao is None:
            return _problema_explicacao_nao_encontrada(elegibilidade_id)
        return _resposta_explicacao(explicacao)

    return roteador
